### PR TITLE
Expose iterator shuffling through Trainer config

### PR DESCRIPTION
**Pull request by nelson-liu**
Not really sure how to test this, any suggestions?

### CODE REVIEW

**Comment by joelgrus** [comment]
hmm, are we shuffling the validation set by default? that's kind of weird?

it's ok if you don't want to test this. if I really wanted to test this, I would probably (in the test file) create a `BasicIterator` subclass that remembers its instances as it iterates over them, then I would run training just as the current `TrainerTest` does (maybe for 10 or 20 epochs) and check that in the the no-shuffle case the instances are are in the same order each epoch and that in the shuffle case they're not. but that might be more work than is worthwhile.

File: allennlp/training/trainer.py [hunk 1]
```diff
@@ -166,6 +166,7 @@ def __init__(self,
                  patience: Optional[int] = None,
                  validation_metric: str = "-loss",
                  validation_iterator: DataIterator = None,
+                 shuffle: bool = True,
```
Comment (joelgrus) [inline allennlp/training/trainer.py]: `Trainer` has a custom from_params, so you'll need to add it there too

Comment (nelson-liu) [reply allennlp/training/trainer.py]: good catch, thanks!

**Comment by nelson-liu** [comment]
> hmm, are we shuffling the validation set by default? that's kind of weird?

yeah, I noticed this too. I'd be in favor of turning off shuffling at validation, since I presume that's what we do at test time? thoughts?

**Comment by joelgrus** [comment]
it looks like we are shuffling at test time too. 😬 

anyway, it appears that we've always been shuffling the validation set, so I wouldn't casually change that behavior. (in theory it shouldn't make a difference, addition is commutative, but who knows)

**Comment by DeNeutoy** [comment]
Can someone remind me why this isn't a constructor arg for the `Iterator`? The way we are passing it here means it's essentially equivalent, as you can't choose to change it whilst training.

**Comment by joelgrus** [comment]
I assumed it was so the same iterator could behave differently on training / dev / test, although that's not possible right now
